examples/ble-gatt-server/src/main/java/no/nordicsemi/android/ble/ble_gatt_server/GattService.kt

```diff
@@ -0,0 +1,299 @@
+                when (intent?.action) {
+                    BluetoothAdapter.ACTION_STATE_CHANGED -> {
+                        val bluetoothState = intent.getIntExtra(
+                                BluetoothAdapter.EXTRA_STATE,
+                                -1
+                        )
+                        when (bluetoothState) {
+                            BluetoothAdapter.STATE_ON -> enableBleServices()
+                            BluetoothAdapter.STATE_OFF -> disableBleServices()
+                        }
+                    }
+                }
+            }
+        }
+        registerReceiver(bluetoothObserver, IntentFilter(BluetoothAdapter.ACTION_STATE_CHANGED))
+
+        // Startup BLE if we have it
+
+        val bluetoothManager = getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
+        if (bluetoothManager.adapter?.isEnabled == true) enableBleServices()
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        disableBleServices()
+    }
+
+    private fun enableBleServices() {
+        serverManager = ServerManager(this)
+        serverManager!!.open()
+
+        bleAdvertiseCallback = BleAdvertiser.Callback()
+
+        val bluetoothManager = getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
+        bluetoothManager.adapter.bluetoothLeAdvertiser?.startAdvertising(
+                BleAdvertiser.settings(),
+                BleAdvertiser.advertiseData(),
+                bleAdvertiseCallback!!
+        )
+    }
+
+    private fun disableBleServices() {
+        if (bleAdvertiseCallback != null) {
+            val bluetoothManager = getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
+            bluetoothManager.adapter.bluetoothLeAdvertiser?.stopAdvertising(bleAdvertiseCallback!!)
+            bleAdvertiseCallback = null
+        }
+
+        serverManager?.close()
+        serverManager = null
+    }
+
+    override fun onBind(intent: Intent?): IBinder? {
+        return DataPlane()
+    }
+
+    /**
+     * Functionality available to clients
+     */
+    inner class DataPlane : Binder() {
+        /**
+         * Change the value of the GATT characteristic that we're publishing
+         */
+        fun setMyCharacteristicValue(value: String) {
+            serverManager?.setMyCharacteristicValue(value)
+        }
+    }
+
+    companion object {
+        private const val TAG = "gatt-service"
+    }
+
+    /*
+     * Manages the entire GATT service, declaring the services and characteristics on offer
+     */
+    private class ServerManager(val context: Context) : BleServerManager(context), ServerObserver {
+
+        private val myGattCharacteristic = sharedCharacteristic(MyServiceProfile.MY_CHARACTERISTIC_UUID,
+                BluetoothGattCharacteristic.PROPERTY_READ // properties
+                        or BluetoothGattCharacteristic.PROPERTY_NOTIFY,
+                BluetoothGattCharacteristic.PERMISSION_READ_ENCRYPTED_MITM,  // permissions
```

Comment:
I had thought that declaring the read to require encryption would cover off notifications also. It seems not though as I'm able to use nRF Connect to connect with the GATT server and receive notifications without having bonded. Any guidance on what else I must do to ensure that notifications are also encrypted?

Reply:
You may set `PERMISSION_WRITE_ENCRYPTED_MITM` on the CCC descriptor. This will trigger bonding when client tries to enable notifications.

Reply:
Ah right. Thanks.

Reply:
Just added a new commit